Endpoint set up measurement nightly test

nebril opened the pull request:
This test:
- Creates a number of k8s pods
- Waits for all deployment pods to be running
- Runs `cilium endpoint list` on both k8s nodes until all pods have
corresponding ready endpoints
- Reports how long did above steps take
- Runs curl from client pods to server pods while restarting cilium pod

part of #2029

nebril committed:
Endpoint set up measurement nightly test

This test:
- Creates 2 deployments
- Waits for all deployment pods to be running
- Runs `cilium endpoint list` on both k8s nodes until all pods have
corresponding ready endpoints
- Reports how long did above steps take

For testing purposes deployments are small (10 pods)

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

ianvernon commented on test/k8sT/Nightly.go:
```diff
@@ -0,0 +1,99 @@
+	log "github.com/sirupsen/logrus"
+)
+
+var _ = Describe("K8sNightlyTest", func() {
```
Change this to be "NightlyK8sEpsMeasurement" - that way we can run ginkgo commands with a specific focus, i.e. `ginkgo --focus="NightlyK8s*" -v -noColor` for K8s nightly tests. Otherwise, this will get run with all Ginkgo tests since it currently is prefixed with "K8s"

Alternatively, we could have a folder called nightly and throw all nightly tests in there?

eloycoto replied:
Not agree here, we should use `K8s` prefix to get the scope, and uses `-skip` flag on the PR test. IMHO

eloycoto replied:
I've just seen that you add Nightly to the GetScope function. @nebril could you change the name of the testSuite to `NightlyK8sEpsMeasurement`?

Regards

nebril replied:
Done

nebril committed:
Add templated k8s manifests to nightly tests

This change generates manifests with Go templates based on two provided
demo manifests. It allows to dynamically change how much endpoints will
be created for a test.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Generate manifest into file

Sending long k8s manifest over ssh command took too much time. File sync
to vagrant is much faster.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Fix duplicate objects in generated k8s manifest

Add html data for server pods in nightly test

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Connectivity nightly tests for endpoints

Simple curl request is performed from each client pod to each
server pod.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

eloycoto commented on test/helpers/manifest_generator.go:
```diff
@@ -0,0 +1,83 @@
+	"text/template"
+)
+
+var manifestBase = "k8sT/manifests"
```
I will move this to `helpers/cons.go`

eloycoto commented on test/helpers/manifest_generator.go:
```diff
@@ -0,0 +1,83 @@
+		return "", 0, err
+	}
+
+	serverTemplateStr, err := ioutil.ReadFile(path.Join(manifestBase, "server.yaml"))
```
If you can move  `server.yaml`, `client.yaml` to `helpers/cons.go` it will be great.

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -0,0 +1,123 @@
+		_, lastServer, err := helpers.GenerateManifestForEndpoints(endpointCount, manifestPath)
+		Expect(err).Should(BeNil())
+
+		vagrantManifestPath := "/vagrant/" + manifestPath
```
`helpers.BasePath` instead of /vagrant/

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -0,0 +1,123 @@
+		vagrantManifestPath := "/vagrant/" + manifestPath
+
+		res := kubectl.Apply(vagrantManifestPath)
+		res.ExpectSuccess(res.GetCmd(), res.GetStdOut(), res.GetStdErr())
```
What about: 

`res.ExpectSuccess(fmt.Sprintf("cmd: %s error: %s", res.GetCmd(), res.CombineOutput())`

Easy to read in the junit output ;-)

eloycoto commented on test/helpers/kubectl.go:
```diff
@@ -460,6 +460,17 @@ func (kub *Kubectl) GetCiliumPodOnNode(namespace string, node string) (string, e
 	return res.Output().String(), nil
 }
 
+// TestConnectivityPodService runs connectivity test from pod to ClusterIP service
```
What about 
```
TestConnectivityPodService runs HTTP connectivity test from pod to ClusterIP service
```

nebril replied:
I left out `HTTP` part because we might want to change the connectivity test itself later. I can put it there, but if we make the connectivity test more sophisticated, we need to reflect this in the comment later

eloycoto reviewed (changes requested):
It looks awesome, a few comments and I need to test locally, but looks good ;-)

eloycoto commented on test/k8sT/Nightly.go:
```diff
@@ -0,0 +1,124 @@
+		if initialized == true {
+			return
+		}
+		logger = log.WithFields(log.Fields{"testName": "K8sServiceTest"})
```
`"testName:"NightlyK8sEpsMeasurement"`

eloycoto reviewed (approved):
Minor comment to fix, but awesome code ;-) 

Example output: 
```
------------------------------                             
• [MEASUREMENT]              
NightlyK8sEpsMeasurement     
/home/ecoto/.go/src/github.com/cilium/cilium/test/k8sT/Nightly.go:124    
  10 endpoint creation       
  /home/ecoto/.go/src/github.com/cilium/cilium/test/k8sT/Nightly.go:123  

  Ran 1 samples:             
  Wait for pods:             
    Fastest Time: 55.568s    
    Slowest Time: 55.568s    
    Average Time: 55.568s ± 0.000s                         
  Endpoint creation:         
    Fastest Time: 13.487s    
    Slowest Time: 13.487s    
    Average Time: 13.487s ± 0.000s                         
------------------------------                             
SSS                          
Ran 1 of 46 Specs in 125.025 seconds                       
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 45 Skipped PASS            
```

nebril committed:
Run connectivity test in ssh client copy

Measure test spec is split from connectivity test spec
Concurrency in connectivity test is configurable

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Test connectivity while cilium is redeployed

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

ianvernon commented on test/helpers/cons.go:
```diff
@@ -37,6 +37,9 @@ const (
 	BasePath = "/vagrant/"
 
+	// ManifestBase tells ginkgo suite where to look for manifests
+	ManifestBase = "k8sT/manifests"
```
Nit: rename this to K8sManifestBase

nebril replied:
Will do.

ianvernon commented on test/helpers/kubectl.go:
```diff
@@ -548,6 +548,18 @@ func (kub *Kubectl) GetCiliumPodOnNode(namespace string, node string) (string, e
 
+// TestConnectivityPodService runs HTTP connectivity test from pod to ClusterIP service
+func (kub *Kubectl) TestConnectivityPodService(from, to string) *CmdRes {
+	kub.logCxt.WithFields(log.Fields{
```
Nit: would be nice if instead of a string for IPs we could pass in actual IP objects. This can be done in a follow-up ticket. We should utilize golang constructs in the tests when we can instead of strings.

nebril replied:
None of arguments passed to this method is ip, `from` is a pod name, `to` is a service name.

ianvernon commented on test/helpers/manifest_generator.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// GenerateManifestForEndpoints generates k8s manifests that will create
+// endpointCount cilium endpoints when applied.
```
Nit: cilium --> Cilium

ianvernon commented on test/helpers/manifest_generator.go:
```diff
@@ -0,0 +1,82 @@
+
+// GenerateManifestForEndpoints generates k8s manifests that will create
+// endpointCount cilium endpoints when applied.
+// 1/3 of endpoints is going to be servers, the rest clients.
```
Nit: change to: "1/3 of the endpoints is a server; the rest are clients."

nebril replied:
will do

ianvernon commented on test/helpers/node.go:
```diff
@@ -116,6 +123,29 @@ func (s *SSHMeta) Exec(cmd string) *CmdRes {
+
+	config := nodes[s.nodeName]
+	if config == nil {
+		log.Errorf("No node %s in imported config", s.nodeName)
```
Nit: No --> no

nebril replied:
will do

ianvernon commented on test/helpers/node.go:
```diff
@@ -116,6 +123,29 @@ func (s *SSHMeta) Exec(cmd string) *CmdRes {
+func (s *SSHMeta) GetCopy() *SSHMeta {
+	nodes, err := ImportSSHconfig(s.rawConfig)
+	if err != nil {
+		log.WithError(err).Error("While importing ssh config for meta copy")
```
Nit: While --> while

nebril replied:
will do
